Added post processing (for reasoning tokens) to pipeline

clefourrier opened the pull request:
Fix #869
Tested on Qwen3-0.6B and AIME25: when you select `--remove-reasoning-tags`, it indeeds removes them before sending the answer to the metric (but the model is quite verbose, so a lot of sentences are not changed because the </think> token never appears.

Edit: updated todos based on observation and Lewis feedback:

- [x] need to add correct parsing for reasoning tags provided by the user
- [x] test on AIME24 instead of 25
```
# Reasoning (/think) and not removing tokens
lighteval vllm 'model_name=Qwen/Qwen3-0.6B,data_parallel_size=4,use_chat_template=True,max_model_length=38912,max_num_batched_tokens=100000,generation_parameters={temperature:0.6,top_p:0.95,top_k:20,min_p:0,presence_penalty:1,max_new_tokens:38912},system_prompt=/think' 'lighteval|aime24_avg|0|0,extended|ifeval|0|0' --save-details --output-dir /fsx/clementine/lighteval/_logs/slurm

## 0.6B
|         Task         |Version|        Metric         |Value |   |Stderr|
|----------------------|-------|-----------------------|-----:|---|-----:|
|extended:ifeval:0     |       |prompt_level_strict_acc|0.3290|±  |0.0202| 
|lighteval:aime24_avg:0|       |math_avg@64            |0.1000|±  |0.0343|

## 1.7B
|         Task         |Version|        Metric         |Value |   |Stderr|
|----------------------|-------|-----------------------|-----:|---|-----:|
|extended:ifeval:0     |       |prompt_level_strict_acc|0.3530|±  |0.0206|
|lighteval:aime24_avg:0|       |math_avg@64            |0.4052|±  |0.0625|
```

```
# Reasoning (/think) and removing tokens
lighteval vllm 'model_name=Qwen/Qwen3-0.6B,data_parallel_size=8,use_chat_template=True,max_num_batched_tokens=100000,generation_parameters={temperature:0.6,top_p:0.95,top_k:20,min_p:0,presence_penalty:1,max_new_tokens:38912},system_prompt=/think' 'lighteval|aime24_avg|0|0,extended|ifeval|0|0' --remove-reasoning-tags --save-details --output-dir /fsx/clementine/lighteval/_logs/slurm

## 0.6B
|         Task         |Version|        Metric         |Value |   |Stderr|
|----------------------|-------|-----------------------|-----:|---|-----:|
|extended:ifeval:0     |       |prompt_level_strict_acc|0.5860|±  |0.0212| # vs 59.2
|lighteval:aime24_avg:0|       |math_avg@64            |0.1104|±  |0.0369| # vs 10.7

## 1.7B
|         Task         |Version|        Metric         |Value |   |Stderr|
|----------------------|-------|-----------------------|-----:|---|-----:|
|extended:ifeval:0     |       |prompt_level_strict_acc|0.6839|±  |0.0200| # vs 72.5
|lighteval:aime24_avg:0|       |math_avg@64            |0.4130|±  |0.0624| # vs 48.3
```

```
# No reasoning
lighteval vllm 'model_name=Qwen/Qwen3-0.6B,data_parallel_size=8,use_chat_template=True,max_num_batched_tokens=100000,generation_parameters={temperature:0.7,top_p:0.80,top_k:20,min_p:0,presence_penalty:1.5,max_new_tokens:32768},system_prompt=/nothink' 'lighteval|aime24_avg|0|0,extended|ifeval|0|0' --save-details --output-dir /fsx/clementine/lighteval/_logs/slurm

## 0.6B
|         Task         |Version|        Metric         |Value |   |Stderr|
|----------------------|-------|-----------------------|-----:|---|-----:|
|extended:ifeval:0     |       |prompt_level_strict_acc|0.3235|±  |0.0201| # vs 54.5
|lighteval:aime24_avg:0|       |math_avg@64            |0.1073|±  |0.0382| # vs 3.4

````
- [x] test on IFEVAL
- [x] add an integration test at the pipeline/main level
- [x] add docs with magistral example

clefourrier commented on src/lighteval/main_custom.py:
```diff
@@ -31,10 +31,10 @@
-HELP_PANNEL_NAME_2 = "Logging Parameters"
-HELP_PANNEL_NAME_3 = "Debug Parameters"
-HELP_PANNEL_NAME_4 = "Modeling Parameters"
+HELP_PANEL_NAME_1 = "Common Parameters"
```
nit, unrelated to the PR

clefourrier commented on src/lighteval/models/model_output.py:
```diff
@@ -105,26 +117,32 @@ class ModelResponse:
-        None  # Log probabilities of the unconditioned model (if applicable)
-    )
+    @property
+    def final_text(self) -> list[str]:
```
Yep good idea to store both in the details

clefourrier replied:
both should be stored as we now store the doc and model response objects as such

lewtun commented:
For posterity, could you also share the command you're using to reproduce the AIME scores in the PR description? 

Note that Qwen3 use different sampling parameters for the `/think` and `/no_think` modes (full set described in #872)

clefourrier commented on src/lighteval/logging/info_loggers.py:
```diff
@@ -170,27 +170,10 @@ class Detail:
-            gold_index (list): Indices of the gold targets among the [`choices`]
-            metrics (dict): Metric name to current example score
-
+            doc (Doc): The [`Doc`] object containing the current example information.
```
Unrelated to PR, incorrect doc was updated

clefourrier commented:
@lewtun Ok, added tests, updated doc, added better management of tokens (better checks to make sure they are valid). I feel like the scores I'm getting for Qwen3, reasoning with tokens removed are not too far.

I'm getting a mismatch for Qwen3, no reasoning, but it's unrelated to the thinking tokens PR so unsure if I should take it into account, (plus the eval is @64 it's terribly slow ~5h for a 2B run with vllm, DP8)

clefourrier commented:
@lewtun merging cause I need it for the current big eval, feel free to add comments here later

lewtun commented:
Thanks @clefourrier ! I think the no-reasoning score mismatch can be attributed to the use of `system_prompt=/nothink'` instead of `system_prompt=/no_think'. LGTM overall!
